File: 04-stack/00084-Largest_Rectangle_in_Histogram/00084-largest_rectangle_in_histogram.cpp
// 84: Largest Rectangle in Histogram
// https://leetcode.com/problems/largest-rectangle-in-histogram/


#include <iostream>
#include <vector>
#include <stack>

using namespace std;


class Solution {
public:
    // SOLUTION
    int largestRectangleArea(vector<int>& heights) {
        stack<int> stack;
        int l = heights.size();
        int result = 0;

        for (int i=0; i<=l; i++) {
            int h = i==l ? 0 : heights[i];

            if (stack.empty() || h>=heights[stack.top()])
                stack.push(i);
            else {
                int tp = stack.top();
                stack.pop();
                result = max(result, heights[tp]*(stack.empty() ? i : i-1-stack.top()));
                i--;
            }
        }

        return result;
    }
};


int main() {
    Solution o;

    // INPUT
    vector<int> heights = {2,1,5,6,2,3};

    // OUTPUT
    auto result = o.largestRectangleArea(heights);
    cout<<result<<endl;

    return 0;
}

File: 04-stack/00084-Largest_Rectangle_in_Histogram/00084-largest_rectangle_in_histogram.py
# 84: Largest Rectangle in Histogram
# https://leetcode.com/problems/largest-rectangle-in-histogram/


class Solution:
    def largestRectangleAre(self, heights: list[int]) -> int:
        stack = []
        l = len(heights)
        result = 0

        i=0
        while i<=l:
            h = 0 if i==l else heights[i]

            if len(stack)==0 or h>=heights[stack[-1]]:
                stack.append(i)
            else:
                tp = stack[-1]
                stack.pop()
                result = max(result, heights[tp]*(i if len(stack)==0 else i-1-stack[-1]))
                i-=1
            i+=1

        return result



if __name__ == "__main__":
    o = Solution()

    # INPUT
    heights: list[int] = [2,1,5,6,2,3]

    # OUTPUT
    result = o.largestRectangleAre(heights)
    print(result)